server/src/main/java/org/elasticsearch/action/admin/indices/stats/CommonStats.java

```diff
@@ -125,42 +133,97 @@ public CommonStats(CommonStatsFlags flags) {
-                    case RequestCache -> requestCache = indexShard.requestCache().stats();
-                    case Recovery -> recoveryStats = indexShard.recoveryStats();
-                    case Bulk -> bulk = indexShard.bulkStats();
+                    case Docs -> stats.docs = indexShard.docStats();
+                    case Store -> stats.store = indexShard.storeStats();
+                    case Indexing -> stats.indexing = indexShard.indexingStats();
+                    case Get -> stats.get = indexShard.getStats();
+                    case Search -> stats.search = indexShard.searchStats(flags.groups());
+                    case Merge -> stats.merge = indexShard.mergeStats();
+                    case Refresh -> stats.refresh = indexShard.refreshStats();
+                    case Flush -> stats.flush = indexShard.flushStats();
+                    case Warmer -> stats.warmer = indexShard.warmerStats();
+                    case QueryCache -> stats.queryCache = indicesQueryCache.getStats(indexShard.shardId());
+                    case FieldData -> stats.fieldData = indexShard.fieldDataStats(flags.fieldDataFields());
+                    case Completion -> stats.completion = indexShard.completionStats(flags.completionDataFields());
+                    case Segments -> stats.segments = indexShard.segmentStats(
+                        flags.includeSegmentFileSizes(),
+                        flags.includeUnloadedSegments()
+                    );
+                    case Translog -> stats.translog = indexShard.translogStats();
+                    case RequestCache -> stats.requestCache = indexShard.requestCache().stats();
+                    case Recovery -> stats.recoveryStats = indexShard.recoveryStats();
+                    case Bulk -> stats.bulk = indexShard.bulkStats();
                     case Shards ->
                         // Setting to 1 because the single IndexShard passed to this method implies 1 shard
-                        shards = new ShardCountStats(1);
+                        stats.shards = new ShardCountStats(1);
+                    case Mappings -> throw new IllegalStateException("Flag: " + flag + " must not be used at shard level");
                     default -> throw new IllegalStateException("Unknown Flag: " + flag);
                 }
             } catch (AlreadyClosedException e) {
                 // shard is closed - no stats is fine
             }
         }
+
+        return stats;
+    }
+
+    /**
+     * Filters the given flags for {@link CommonStatsFlags#INDEX_LEVEL} flags and calculates the corresponding statistics.
+     */
+    public static CommonStats getIndexLevelStats(final IndexService indexService, CommonStatsFlags flags) {
+        // Filter shard level flags
+        CommonStatsFlags filteredFlags = flags.clone();
+        Arrays.stream(filteredFlags.getFlags()).forEach(x -> filteredFlags.set(x, CommonStatsFlags.INDEX_LEVEL.isSet(x)));
```

Comment:
We do the filtering because of the `_all` case right? Outside of that case it seems like we should just throw when validating the original stats request and an illegal combination of flags and level is set?

Reply:
Do you mean the REST stats request? That one has all the flags indeed, and only when we reach these two functions (`getShardLevelStats` and `getIndexLevelStats`), we filter the shard-level and the index-level flags. I do not believe we can validate it at the original stats request; because it's not wrong even if it has mappings as a flag and level is shard, it will show the mapping for the node level.

Reply:
Simplified method. Please see new commit.